### PR TITLE
Pipeline creation using GitHub Actions

### PR DESCRIPTION
**Pull request by sujatadhl**
Pipeline to push code base from GitHub repository to S3 bucket and 

Configuring EC2 instance using ansible and automating it through pipeline

### CODE REVIEW

File: ansible/roles/nginx/tasks/main.yml [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+  - name: Install nginx
+    ansible.builtin.apt:
+      name: nginx
+      state: latest 
+
+  - name: Change Nginx Configuration file   
+    ansible.builtin.template:
+      src: nginx.conf.j2
+      dest: /etc/nginx/sites-available/default
```
Comment (milandangol) [inline ansible/roles/nginx/tasks/main.yml]: please use variable

File: ansible/roles/nginx/tasks/main.yml [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+  - name: Install python
```
Comment (milandangol) [inline ansible/roles/nginx/tasks/main.yml]: lets not use apt twice and use loop for it

File: .github/workflows/main.yml [hunk 3]
```diff
@@ -0,0 +1,81 @@
+name: Upload to S3
+on:
+    push:
+
```
Comment (milandangol) [inline .github/workflows/main.yml:9]: please add on pull request as well

File: .github/workflows/main.yml [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+            cd terraform
+            terraform apply -var-file="var.tfvars" -auto-approve
+         
+
+    add-to-s3:
+      needs: create-resources
+      runs-on: ubuntu-latest
+      steps:
+        - name: Checkout code
+          uses: actions/checkout@v3
```
Comment (milandangol) [inline .github/workflows/main.yml:47]: why different versions

File: .github/workflows/main.yml [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+          
+        - name: Configure aws credentials
+          uses: aws-actions/configure-aws-credentials@v1
+          with:
+            aws-access-key-id: ${{ secrets.AWS_ACCESS_KEY_ID }}
```
Comment (milandangol) [inline .github/workflows/main.yml]: why use this when you have define environment

File: ansible/roles/nginx/templates/nginx.conf.j2 [hunk 6]
```diff
@@ -0,0 +1,8 @@
+server {
```
Comment (milandangol) [inline ansible/roles/nginx/templates/nginx.conf.j2:1]: if this file is static please use `files` folder

File: ansible/roles/node/tasks/main.yml [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+      - nodejs
+      - npm   
+  
+  - name: Make folder in EC2 instance
+    file:
+        path: /home/ubuntu/
+        state: directory
+        mode: '0755'
+       
+  - name: Copy React-app folder from S3 to EC2
```
Comment (milandangol) [inline ansible/roles/node/tasks/main.yml:13]: lets not use command

**Review by milandangol — CHANGES_REQUESTED** [review]

File: ansible/roles/nginx/tasks/main.yml (hunk 1 above)
Comment (sujatadhl) [reply ansible/roles/nginx/tasks/main.yml]: used variables for path

File: ansible/roles/nginx/tasks/main.yml (hunk 2 above)
Comment (sujatadhl) [reply ansible/roles/nginx/tasks/main.yml]: installed all using loop

File: .github/workflows/main.yml (hunk 3 above)
Comment (sujatadhl) [reply .github/workflows/main.yml:9]: added pull request

File: .github/workflows/main.yml (hunk 4 above)
Comment (sujatadhl) [reply .github/workflows/main.yml:47]: maintained same version

File: .github/workflows/main.yml [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+
+        - name: configure aws credentials
+          uses: aws-actions/configure-aws-credentials@v4
+          with:
+            aws-access-key-id: ${{ env.AWS_ACCESS_KEY_ID }}
+            aws-secret-access-key: ${{ env.AWS_SECRET_ACCESS_KEY }}
+            aws-session-token: ${{ env.AWS_SESSION_TOKEN }}
+            aws-region:  ${{ env.AWS_DEFAULT_REGION }}
+
+        - name: Install Ansible
```
Comment (milandangol) [inline .github/workflows/main.yml]: lets use requirement.txt

File: ansible/roles/nginx/tasks/main.yml [hunk 9]
```diff
@@ -0,0 +1,29 @@
+  - name: Update  apt packages
+    ansible.builtin.apt:
+      update_cache: yes
+
+  - name: Install nginx
+    ansible.builtin.apt:
+      name: "{{ item }}"
+      state: latest 
+    loop:
```
Comment (milandangol) [inline ansible/roles/nginx/tasks/main.yml]: lets use varibale

File: ansible/roles/nginx/tasks/main.yml [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+      name: "{{ item }}"
+      state: latest 
+    loop: 
+      - nginx
+      - python
+      - awscli
+      
+  - name: Change Nginx Configuration file   
+    ansible.builtin.template:
+      src: files/nginx.conf.j2
```
Comment (milandangol) [inline ansible/roles/nginx/tasks/main.yml]: template should be on template folder

File: ansible/roles/node/tasks/main.yml [hunk 11]
```diff
@@ -0,0 +1,32 @@
+  - name: Install nodejs and npm
+    ansible.builtin.apt:
+      name: "{{ item  }}"
+      state: latest
+    loop:
```
Comment (milandangol) [inline ansible/roles/node/tasks/main.yml]: use variable

**Review by milandangol — CHANGES_REQUESTED** [review]

File: .github/workflows/main.yml (hunk 5 above)
Comment (sujatadhl) [reply .github/workflows/main.yml]: passes credentials using environment

File: ansible/roles/node/tasks/main.yml (hunk 7 above)
Comment (sujatadhl) [reply ansible/roles/node/tasks/main.yml:13]: made changes

File: ansible/roles/nginx/tasks/main.yml (hunk 9 above)
Comment (sujatadhl) [reply ansible/roles/nginx/tasks/main.yml]: passed values in loop through variables

File: ansible/roles/nginx/tasks/main.yml (hunk 10 above)
Comment (sujatadhl) [reply ansible/roles/nginx/tasks/main.yml]: for static file, created a files folder

File: ansible/roles/node/tasks/main.yml (hunk 11 above)
Comment (sujatadhl) [reply ansible/roles/node/tasks/main.yml]: used variables in loop

File: .github/workflows/main.yml [hunk 12]
```diff
@@ -0,0 +1,88 @@
+name: Upload to S3
+on:
+    push:
+      branches:
+        - git-actions
+    pull_request:
```
Comment (milandangol) [inline .github/workflows/main.yml:6]: make it main for pull request

File: ansible/playbook.yml [hunk 13]
```diff
@@ -0,0 +1,12 @@
+---
+- hosts: nginx
+  become: true
+  gather_facts: true
```
Comment (milandangol) [inline ansible/playbook.yml]: No need to use gather facts

**Review by milandangol — APPROVED** [review]

**Pull request merged**